### ISSUE TITLE
Parallel Cache Failure

### ISSUE THREAD

**Issue by blarghmatey** [saltstack/salt#42706]
### Description of Issue/Question
Any time that I try to use the new parallel state capability the states always return False with a comment that says `Parallel cache failure` despite completing successfully.

### Setup
Any state that has `parallel: True`

### Steps to Reproduce Issue
Run a state that has an entry with `parallel: True`

### Versions Report
```
    Salt Version:
               Salt: 2017.7.0
     
    Dependency Versions:
               cffi: Not Installed
           cherrypy: Not Installed
           dateutil: 2.4.2
          docker-py: Not Installed
              gitdb: Not Installed
          gitpython: Not Installed
              ioflo: Not Installed
             Jinja2: 2.8
            libgit2: Not Installed
            libnacl: Not Installed
           M2Crypto: Not Installed
               Mako: 1.0.3
       msgpack-pure: Not Installed
     msgpack-python: 0.4.6
       mysql-python: 0.7.2.None
          pycparser: Not Installed
           pycrypto: 2.6.1
       pycryptodome: Not Installed
             pygit2: Not Installed
             Python: 2.7.12 (default, Nov 19 2016, 06:48:10)
       python-gnupg: Not Installed
             PyYAML: 3.11
              PyZMQ: 15.2.0
               RAET: Not Installed
              smmap: Not Installed
            timelib: Not Installed
            Tornado: 4.2.1
                ZMQ: 4.1.4
     
    System Versions:
               dist: Ubuntu 16.04 xenial
             locale: UTF-8
            machine: x86_64
            release: 4.4.0-1020-aws
             system: Linux
            version: Ubuntu 16.04 xenial
```

**Comment by garethgreenaway** [comment]
@blarghmatey Thanks for the report.  Can you provide an example of one of the states that is failing using parallel: True?

**Comment by blarghmatey** [comment]
@garethgreenaway this is the relevant portion of one of the state runs that is exhibiting this behavior:
```salt
{% for service in salt.pillar.get('backups:enabled', []) %}
{% if service.get('pkgs') %}
install_packages_for_{{ service.title }}_backup:
  pkg.installed:
    - pkgs: {{ service.pkgs }}
{% endif %}
{% endfor %}

{% for service in salt.pillar.get('backups:enabled', []) %}
run_backup_for_{{ service.title }}:
  file.managed:
    - name: /backups/{{service.title}}_backup.sh
    - source: salt://backups/templates/backup_{{ service.name }}.sh
    - template: jinja
    - context:
        settings: {{ service.settings }}
    - parallel: True
  cmd.script:
    - name: salt://backups/templates/backup_{{ service.name }}.sh
    - template: jinja
    - context:
        settings: {{ service.settings }}
    - parallel: True
    - require_in:
        - event: wait_for_backups_to_complete
{% endfor %}

wait_for_backups_to_complete:
  file.touch:
    - name: /backups/backup_complete
```

**Comment by jubrad** [comment]
I am also seeing this. I was able to consistently reproduce in a centos 7 docker container with `salt-master 2017.7.0 (Nitrogen)`

I believe the issue is coming from a `/` char in the tag name generated from state preventing the cache file creation here:
https://github.com/saltstack/salt/blob/333d86d96c87b615e01bd182e5c13ae8180f6904/salt/state.py#L1709


It's probably best to use an os encoded version of the tag name or a uuid for the state cache file.



```bash
[root@5f3ce76caa67 /]# salt-call state.apply test_parallel
[ERROR   ] An un-handled exception from the multiprocessing process 'MultiprocessingProcess-1' was caught:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/utils/process.py", line 647, in _run
    return self._original_run()
  File "/usr/lib64/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python2.7/site-packages/salt/state.py", line 1705, in _call_parallel_target
    with salt.utils.fopen(tfile, 'wb+') as fp_:
  File "/usr/lib/python2.7/site-packages/salt/utils/__init__.py", line 1352, in fopen
    fhandle = open(*args, **kwargs)  # pylint: disable=resource-leakage
IOError: [Errno 2] No such file or directory: '/var/cache/salt/minion/20170817135229422073/cmd_|-m1_|-date && touch /test_|-run'
Process MultiprocessingProcess-1:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python2.7/site-packages/salt/utils/process.py", line 647, in _run
    return self._original_run()
  File "/usr/lib64/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python2.7/site-packages/salt/state.py", line 1705, in _call_parallel_target
    with salt.utils.fopen(tfile, 'wb+') as fp_:
  File "/usr/lib/python2.7/site-packages/salt/utils/__init__.py", line 1352, in fopen
    fhandle = open(*args, **kwargs)  # pylint: disable=resource-leakage
IOError: [Errno 2] No such file or directory: '/var/cache/salt/minion/20170817135229422073/cmd_|-m1_|-date && touch /test_|-run' 
local:
----------
          ID: m1
    Function: cmd.run
        Name: date && touch /test
      Result: False
     Comment: Parallel cache failure
     Started: 13:52:30.965968
    Duration: 3.43 ms
     Changes:
----------
          ID: m2
    Function: cmd.run
        Name: sleep 2
      Result: True
     Comment: Command "sleep 2" run
     Started: 13:52:30.969850
    Duration: 2.567 ms
     Changes:
              ----------
              pid:
                  3037
              retcode:
                  0
              stderr:
              stdout:

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    1
------------
Total states run:     2
Total run time:   5.997 ms
```

**Comment by damon-atkins** [comment]
The issue here will be the `/` in the filename.
i.e. `cmd_|-m1_|-date && touch` a directory `containing test_|-run`
vs
`cmd_|-m1_|-date && touch /test_|-run`

Windows replaces bad chars with _ . I suspect url encoding will not fix this as `/` will be left alone.

**Comment by iavael** [comment]
Recenty tried to set "parallel: True" in some states and saw this error. Salt version is 2017.7.2 (both master and minion)

**Comment by ferringb** [comment]
This is fixed via salt 2017.7.3, specifically https://github.com/saltstack/salt/commit/1dcf167bb7befa0e1ebf8c7256437318961f0351 .
